src/samplers/tcp/bpf.c

```diff
@@ -269,3 +272,31 @@ int trace_rto(struct pt_regs *ctx, struct sock *sk)
     add_value(rto.lookup(&loc), 1);
     return 0;
 }
+
+// Run on incoming package validation
+int trace_validate_incoming(struct pt_regs *ctx, struct sock *sk, struct sk_buff *skb) {
+    if (sk == NULL || skb == NULL)
+        return 0;
+
+    struct tcp_sock *tp = tcp_sk(sk);
+    u32 seq = TCP_SKB_CB(skb)->seq;
+    u32 end_seq = TCP_SKB_CB(skb)->end_seq;
+
+    // Segment sequence before the expected one
+    // which means this was a duplicated segment
+    if ((end_seq - tp->rcv_wup) < 0) {
```

Comment:
I feel the starting sequence number TCP_SKB_CB(skb)->seq and the expected sequence number is tp->rcv_nxt is more straightforward to compare

if TCP_SKB_CB(skb)->seq is smaller than tp->rcv_nxt, it's confirmed a duplicate.

Reply:
'rcv_wup' records the rcv_nxt on last window update sent, comparing it to the end_seq of this segment does not seem correct to me.

Reply:
I'm basing these on this function that is used to validate the segment sequence:
https://elixir.bootlin.com/linux/v4.2/source/net/ipv4/tcp_input.c#L3902

From the comments it states:
```
Also, controls (RST is main one) are accepted using RCV.WUP instead 
of RCV.NXT. Peer still did not advance his SND.UNA when we 
delayed ACK, so that hisSND.UNA<=ourRCV.WUP.
```
So to validate it seems better if we use RCV.WUP. But maybe for probing duplicates we don't care about this and should use RCV.NXT anyways?

Reply:
I think what 

```
!before(end_seq, tp->rcv_wup) &&
 !after(seq, tp->rcv_nxt + tcp_receive_window(tp))
```
checks eventually is that whether the sender's segment [seq, end_seq] has any overlays with the expected window at all. 
The first check finds out if the end_seq is even before the previous rcv_nxt sent, the second check finds out if the start seq is even after the largest possible seq (which is rcv_nxt + window size).

For us we simply wants to check if the segment is a duplicate, and checking end_seq - tp->rcv_wup can't help, for example, when (end_seq - tp->rcv_wup) <0, we can confirm that it's a duplicate, but when (end_seq - tp->rcv_wup) > 0, it can still be a duplicate, as long as seq < tp->rcv_nxt. 

So basically to check duplicates, we should be looking seq and tp->rct_nxt.  What do you think?

Reply:
Yes, this makes sense. I made the changes to use RCV.NXT instead.